be/src/storage/rowset/column_reader.cpp

```diff
@@ -269,27 +278,29 @@ Status ColumnReader::_init(ColumnMetaPB* meta) {
             _sub_readers->reserve(3);
 
             // keys
-            auto res = ColumnReader::create(meta->mutable_children_columns(0), _segment);
+            auto res = ColumnReader::create(meta->mutable_children_columns(0), _segment, nullptr);
             RETURN_IF_ERROR(res);
             _sub_readers->emplace_back(std::move(res).value());
 
             // values
-            res = ColumnReader::create(meta->mutable_children_columns(1), _segment);
+            res = ColumnReader::create(meta->mutable_children_columns(1), _segment, nullptr);
             RETURN_IF_ERROR(res);
             _sub_readers->emplace_back(std::move(res).value());
 
             // offsets
-            res = ColumnReader::create(meta->mutable_children_columns(2), _segment);
+            res = ColumnReader::create(meta->mutable_children_columns(2), _segment, nullptr);
             RETURN_IF_ERROR(res);
             _sub_readers->emplace_back(std::move(res).value());
         }
         return Status::OK();
     } else if (_column_type == LogicalType::TYPE_STRUCT) {
         _sub_readers = std::make_unique<SubReaderList>();
         for (int i = 0; i < meta->children_columns_size(); ++i) {
-            auto res = ColumnReader::create(meta->mutable_children_columns(i), _segment);
+            auto sub_column = (column != nullptr) ? column->subcolumn_ptr(i) : nullptr;
```

Comment:
`meta` is from segment footer.
`column` is from latest schema.
`i` may be not consistent?

Reply:
> `meta` is from segment footer. `column` is from latest schema. `i` may be not consistent?

When initializing the column reader, the `column` is the one corresponding to the tablet schema of this segment, not the `column` from the latest schema.